---
title: "Allele Frequency_RfMS_SNP"
author: "Seungmo Kim"
date: "9/18/2017"
output:
  html_document:
    keep_md: true
---


#for plotting SNPs of S95 & S96 BSA plants
#for clean-up
```{r}
library(tidyverse)
library(ggplot2)
library(caTools) #Could also try using IRanges or zoo packages
#install.packages("Biostrings")
#library(Biostrings)
library(plyr)
library(reshape2)
```

##### read .vcf file and give it column names from header
```{r}
data <- read.delim("S95S96_filtered.vcf.recode.vcf", stringsAsFactors = FALSE, comment.char = "#", header=FALSE)

colnames(data) <- unlist(
  strsplit(
    sub("#","",system("zgrep '^#[A-Z,a-z]' S95S96_filtered.vcf.recode.vcf", intern = TRUE)),
    split = "\t")
  )

head(data)
tail(data)
names(data)
```
##### to break up the data in the final two columns and give them names( 9 columns, not 7)
```{r}
data$S95[is.na(data$S95)] <- "NA:NA:NA:NA:NA:NA:NA:NA:NA"
 
S95.tmp <- matrix(
  unlist(strsplit(data$S95,split = ":")),
  nrow=nrow(data),
  byrow=TRUE
  )
head(S95.tmp)

colnames(S95.tmp) <- paste("S95",c("gt","gtQ","read.depth", "dpr","ref.depth","ref.qual","alt.depth","alt.qual","gt.like"),sep="_")


data$S96[is.na(data$S96)] <- "NA:NA:NA:NA:NA:NA:NA:NA:NA"
 
S96.tmp <- matrix(
  unlist(strsplit(data$S96,split = ":")),
  nrow=nrow(data),
  byrow=TRUE
  )
head(S96.tmp)

colnames(S96.tmp) <- paste("S96",c("gt","gtQ","read.depth", "dpr","ref.depth","ref.qual","alt.depth","alt.qual","gt.like"),sep="_")

#gt: the most probable genotype. 1/1 is homozygous alternate; 0/0 is homozygous reference; 0/1 is heterozygous.
#gtQ: Genotype Quality, the Phred-scaled marginal probability of the called genotype 
#read_depth
#dpr: Number of observation for each allele 
#ref.depth: Reference allele observation count 
#ref.qual: Sum of quality of the reference observations 
#alt.depth: Alternate allele observation count 
#alt.qual: Sum of quality of the alternate observations 
#gt.like: Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy 

data <- cbind(data,S95.tmp,S96.tmp,stringsAsFactors=FALSE)
summary(data)
```

##### conver the columns into numeric
```{r}
head(data)

#"gt","gtQ","read_depth", #"dpr","ref.depth","ref.qual","alt.depth","alt.qual","gt.like"

data[,c("S95_gtQ", "S95_read.depth","S95_ref.depth","S95_ref.qual","S95_alt.depth","S95_alt.qual","S96_gtQ", "S96_read.depth","S96_ref.depth","S96_ref.qual","S96_alt.depth","S96_alt.qual")] <- apply(data[,c("S95_gtQ","S95_read.depth","S95_ref.depth","S95_ref.qual","S95_alt.depth","S95_alt.qual", "S96_read.depth","S96_gtQ","S96_ref.depth","S96_ref.qual","S96_alt.depth","S96_alt.qual")], 2, as.numeric)
head(data)
summary(data)
```

##### add % of alternate alleles
```{r}
data <- data %>%
   mutate(S95_percent.alt=S95_alt.depth/S95_read.depth*100, S96_percent.alt=S96_alt.depth/S96_read.depth*100, QUAL_converted =(10)*log10(QUAL))
head(data)
```

# check out genotype quality values before filtering
```{r}
qplot(data$QUAL, geom="histogram") 
qplot(data$QUAL, geom="histogram") + xlim(0,10000)
summary(data$QUAL)

qplot(data$QUAL_converted, geom="histogram") 
summary(data$QUAL_converted)
```


#S95 & S96 coverage
```{r}
summary(data$S96_read.depth)
qplot(data$S96_read.depth,geom="histogram") + scale_x_log10()
qplot(data$S96_read.depth,geom="histogram") + xlim(0,100)

summary(data$S95_read.depth)
qplot(data$S95_read.depth,geom="histogram") + scale_x_log10()
qplot(data$S95_read.depth,geom="histogram") + xlim(0,100)
```


#Compare the alt. allele frequency between S96 & S95
```{r}
chr_names <- read.csv("chromosome_names.txt", header=FALSE)
chr_names$V1 <- as.character(chr_names$V1)
chr_names <- unlist(chr_names)

data1 <- data %>% 
          select(CHROM, POS, S96_alt.depth, S95_alt.depth) 

# return a graph by chromosome
for (chrom in chr_names) {
   data2 <- data1 %>% filter(CHROM == chrom)
   ggplot(data=data2) +
      geom_smooth(aes(x=POS,y=S96_alt.depth), color="blue") +
      geom_smooth(aes(x=POS,y=S95_alt.depth), color="red") +
      ggtitle(paste0("alt. allele frequency in ",chrom)) + 
      xlab("POS") + ylab("alt. allele freq.")
   ggsave(filename = paste0(chrom,".plot.png"),device = "png",path = "plots")
}
```




